File: c/binarytree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct NodeType {
   struct NodeType* left;
   struct NodeType* right;
   int   data;
}Node;

void InsertBinaryTree(Node** top, Node* new) {
   if ((*top) != NULL) {
      printf("currTop->data = %d : new->data = %d\n", (*top)->data, new->data);
   } else {
      printf("currTop = NULL : new->data = %d\n",  new->data);
   }
   if ((*top) == NULL) {
      *top = new;
      //top->left = NULL;
      //top->right = NULL;
   } else if ((*top)->data > new->data) { // insert to the left
      printf("go left, ");
      InsertBinaryTree(&(*top)->left, new);
   } else if ((*top)->data <= new->data) {
      printf("go right, ");
      InsertBinaryTree(&(*top)->right, new);
   }
}

void PrintBinaryTreeInfix(Node *top) {
   if (top != NULL) {
      if (top->left != NULL) {
         PrintBinaryTreeInfix(top->left);
      }
      printf(" %d ", top->data);

      if (top->right != NULL) {
            PrintBinaryTreeInfix(top->right);
      }
   }
}

void PrintBinaryTreePrefix(Node *top) {
   if (top != NULL) {

      printf(" %d ", top->data);

      if (top->left != NULL) {
         PrintBinaryTreePrefix(top->left);
      }

      if (top->right != NULL) {
            PrintBinaryTreePrefix(top->right);
      }
   }
}

void PrintBinaryTreePostfix(Node *top) {
   if (top != NULL) {
      if (top->left != NULL) {
         PrintBinaryTreePrefix(top->left);
      }

      if (top->right != NULL) {
            PrintBinaryTreePrefix(top->right);
      }

      printf(" %d ", top->data);
   }
}

int main() {
   Node* root = NULL;
   Node* tmp = NULL;
   
   int dummy[] = { 6, 9, 7, 3, 1, 4, 5, 2 };

   int nDummies = 8;

   for (int i=0; i<nDummies; i++) {
      tmp = (Node*)malloc(sizeof(Node));
      if (tmp != NULL) {
         tmp->data = dummy[i];
//         printf("tmp->data = %d\n", tmp->data);
         tmp->left = NULL;
         tmp->right = NULL;
         InsertBinaryTree(&root, tmp);
      } else {
         printf("Error: out of memory!\n");
         exit (-1);
      }
   }

   printf("Print Binary Tree - Prefix:");
   PrintBinaryTreePrefix(root);
   printf("\n");

   printf("Print Binary Tree - Infix:");
   PrintBinaryTreeInfix(root);
   printf("\n");

   printf("Print Binary Tree - Postfix:");
   PrintBinaryTreePostfix(root);
   printf("\n");
}

File: python/cs110-binary_tree/binaryTree.py
from collections import deque

'''
class Node:
    def __init__(self, data=None, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    def print(self):
        print(self.data)
'''

# since every node is a tree, so we need only class Tree
class Tree:
    def __init__(self, data=None, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    def insert(self, data):
        if self.data is None:
            self.data = data
        else:
            if data < self.data:
                if self.left:
                    self.left.insert(data)
                else:
                    self.left = Tree(data)
            elif data > self.data:
                if self.right:
                    self.right.insert(data)
                else:
                    self.right = Tree(data)

    def print(self):
        if self.left:
            self.left.print()

        print(self.data)

        if self.right:
            self.right.print()

    '''
    def print(self):
        curr = self.root

        level = 0

        q = []
        q.append(self.root)

        while len(q):
            print("level %d" % level)
            for i in len(q):
                curr = q[0]
                del q[0]
                print("%d " % curr.data, end='') 
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)
            level += 1
    '''

t = Tree(3)
t.print()

t.insert(1)
t.insert(2)
t.insert(4)
t.print()
